prototype_source/fx_graph_mode_ptq_static.rst

```diff
@@ -1,64 +1,66 @@
-  
-Advantages of FX Graph Mode Quantization are: 
-  
-- Simple quantization flow, minimal manual steps  
-- Unlocks the possibility of doing higher level optimizations like automatic precision selection  
-  
-2. Define Helper Functions and Prepare Dataset    
-----------------------------------------------    
-  
-We’ll start by doing the necessary imports, defining some helper functions and prepare the data.  
-These steps are identitcal to `Static Quantization with Eager Mode in PyTorch <https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html>`_.       
+  from torch.ao.quantization import QConfigMapping
+  float_model.eval()
+  qconfig = get_default_qconfig("fbgemm")
+  qconfig_mapping = QConfigMapping.set_global(qconfig)
+  def calibrate(model, data_loader):
+      model.eval()
+      with torch.no_grad():
+          for image, target in data_loader:
+              model(image)
+  example_inputs = (next(iter(data_loader))[0]) # get an example input
+  prepared_model = prepare_fx(float_model, qconfig_mapping, example_inputs)  # fuse modules and insert observers
+  calibrate(prepared_model, data_loader_test)  # run calibration on sample data
+  quantized_model = convert_fx(prepared_model)  # convert the calibrated model to a quantized model
+
+
+
+1. Motivation of FX Graph Mode Quantization
+-------------------------------------------
+
+Currently PyTorch only has eager mode quantization: `Static Quantization with Eager Mode in PyTorch <https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html>`_.
+
+We can see there are multiple manual steps involved in the process, including:
+
+- Explicitly quantize and dequantize activations, this is time consuming when floating point and quantized operations are mixed in a model.
+- Explicitly fuse modules, this requires manually identifying the sequence of convolutions, batch norms and relus and other fusion patterns.
+- Special handling is needed for pytorch tensor operations (like add, concat etc.)
+- Functionals did not have first class support (functional.conv2d and functional.linear would not get quantized)
+
+Most of these required modifications comes from the underlying limitations of eager mode quantization. Eager mode works in module level since it can not inspect the code that is actually run (in the forward function), quantization is achieved by module swapping, and we don’t know how the modules are used in forward function in eager mode, so it requires users to insert QuantStub and DeQuantStub manually to mark the points they want to quantize or dequantize.
```

Comment:
```suggestion
Most of these required modifications come from the underlying limitations of eager mode quantization. Eager mode works in module level since it can not inspect the code that is actually run (in the forward function), quantization is achieved by module swapping, and we don’t know how the modules are used in forward function in eager mode, so it requires users to insert QuantStub and DeQuantStub manually to mark the points they want to quantize or dequantize.
```